### PR TITLE
fix: shared queue requires aligned buffer

### PR DESCRIPTION
**Pull request by bartlomieju**
This PR fixes problem described in #2814, namely: `shared_queue` requires buffers that are aligned to 4-bytes so that `dispatch_minimal` can work (it uses Int32Array, which requires that `byteOffset` of underlying buffer is a multiple of 4)

CC @ry

### CODE REVIEW

File: js/dispatch_json.ts [hunk 1]
```diff
@@ -30,7 +30,10 @@ function nextPromiseId(): number {
 }
 
 function decode(ui8: Uint8Array): JsonResponse {
-  const s = new TextDecoder().decode(ui8);
+  let s = new TextDecoder().decode(ui8);
+  // TODO: this is make-shift solution
+  const closingBracket = s.lastIndexOf("}");
+  s = s.slice(0, closingBracket + 1);
```
Comment (ry) [inline js/dispatch_json.ts]: This seems odd / sub-optimal. What characters are you padding with? I suggest using space characters - then you shouldn't need to slice.

Comment (bartlomieju) [reply js/dispatch_json.ts]: That's so obvious... Thanks!

File: js/dispatch_json.ts [hunk 2]
```diff
@@ -30,11 +30,10 @@ function nextPromiseId(): number {
...
 
 function decode(ui8: Uint8Array): JsonResponse {
+  // Received JSON string may be padded with "space" characters
+  // at the end, so make sure to trim it before parsing.
   let s = new TextDecoder().decode(ui8);
-  // TODO: this is make-shift solution
-  const closingBracket = s.lastIndexOf("}");
-  s = s.slice(0, closingBracket + 1);
-  return JSON.parse(s) as JsonResponse;
+  return JSON.parse(s.trimEnd()) as JsonResponse;
```
Comment (nayeemrmn) [inline js/dispatch_json.ts]: Is there any need to trim it?

Comment (bartlomieju) [reply js/dispatch_json.ts]: Yes, take a look at above comment

File: cli/ops/dispatch_json.rs [hunk 3]
```diff
@@ -43,9 +43,11 @@ fn serialize_result(
   let len = vec.len();
   let modulo = len % 4;
 
+  // Pad string with "space" character. Required by shared_queue.
   if modulo != 0 {
     let new_len = len + (4 - modulo);
-    vec.resize(new_len, 0);
+    let space_byte = 32;
+    vec.resize(new_len, space_byte);
```
Comment (ry) [inline cli/ops/dispatch_json.rs]: Try something like this
```rust
vec.resize(new_len, ' ' as u8);
```

File: js/dispatch_json.ts (hunk 2 above)
Comment (nayeemrmn) [reply js/dispatch_json.ts]: `JSON.parse()` will just ignore whitespace, won't it?

File: cli/ops/dispatch_json.rs (hunk 3 above)
Comment (bartlomieju) [reply cli/ops/dispatch_json.rs]: Awesome, thanks!

File: js/dispatch_json.ts (hunk 2 above)
Comment (bartlomieju) [reply js/dispatch_json.ts]: @nayeemrmn good catch, it will. Removed the trim, thanks :+1:

File: js/dispatch_json.ts [hunk 4]
```diff
@@ -30,7 +30,7 @@ function nextPromiseId(): number {
 }
 
 function decode(ui8: Uint8Array): JsonResponse {
-  const s = new TextDecoder().decode(ui8);
+  let s = new TextDecoder().decode(ui8);
```
Comment (nayeemrmn) [inline js/dispatch_json.ts]: Now revert this to `const s` and this file goes untouched :)

Comment (bartlomieju) [reply js/dispatch_json.ts]: Doh, my bad, thanks again

File: cli/ops/dispatch_json.rs [hunk 5]
```diff
@@ -37,7 +37,19 @@ fn serialize_result(
     Ok(v) => json!({ "ok": v, "promiseId": promise_id }),
     Err(err) => json!({ "err": json_err(err), "promiseId": promise_id }),
   };
-  let vec = serde_json::to_vec(&value).unwrap();
+  let mut vec = serde_json::to_vec(&value).unwrap();
+
+  debug!("JSON response pre-align, len={}", vec.len());
+  let len = vec.len();
+  let modulo = len % 4;
```
Comment (piscisaureus) [inline cli/ops/dispatch_json.rs]: This works, no comment about that.
That said, this whole section can be replaced by a one-liner:
```rust
vec.resize((vec.len() + 3usize) & !4usize, b' ');
```

Comment (ry) [reply cli/ops/dispatch_json.rs]: Please leave a comment with that--^

**Pull request merged**